### PR TITLE
chore: add min version in TAV matrix

### PR DESCRIPTION
**Pull request by david-luna**
Tweak `tav.json` file and `tav.yml` and `tav-command.yml` actions so we can specify which is the min version where the module should be tested. The matrix generated contains only one row where each element contains `module node_version` in a single place.

set to draft to test it and discuss with the team

### Checklist

- [x] Implement code
- [ ] Add tests
- [ ] Update TypeScript typings
- [ ] Update documentation
- [x] Add CHANGELOG.asciidoc entry
- [x] Commit message follows [commit guidelines](https://github.com/elastic/apm-agent-nodejs/blob/main/CONTRIBUTING.md#commit-message-guidelines)

### CODE REVIEW

**Review by david-luna — COMMENTED** [review]
/test tav

**Review by david-luna — COMMENTED** [review]
/test tav

**Comment by david-luna** [comment]
The matrix workflow shows the values of each dimension for each run. In our case dimensions are
- `nodeAndVersion`: which contains the string to use in the TAV command (eg. hapi 16)
- `count`: the total number of elements in the matrix


When looking at the workflow, the list of runs differs a little bit of what we are used to
Screenshot 2023-08-01 at 16 57 09

**Review by david-luna — COMMENTED** [review]
/test tav fastify

**Review by david-luna — COMMENTED** [review]
/test tav fastify 10

**Comment by david-luna** [comment]
> /test tav fastify 10

this doe not run anything since min node version set is higher than 10

**Review by david-luna — COMMENTED** [review]
/test tav unknown

**Review by david-luna — COMMENTED** [review]
/test tav unknown

**Comment by david-luna** [comment]
> /test tav unknown

this fails because the package name is not in listed in `tav.json`

**Review by david-luna — COMMENTED** [review]
/test tav

File: .ci/tav.json [hunk 1]
```diff
@@ -1,38 +1,38 @@
...
-    "pg",
-    "redis",
-    "restify",
-    "tedious",
-    "undici",
-    "ws",
-    "@koa/router,koa-router",
-    "handlebars,pug",
-    "bluebird,got"
+    {"name": "@apollo/server", "minVersion": 18 },
```
Comment (david-luna) [inline .ci/tav.json:4]: Note for reviewer: to run less tests we add `minVersion` to 18. This shall be updated with the proper versions if the PR is okay to merge

File: .ci/tav.json
Comment (trentm) [inline .ci/tav.json:1]: Note that workflows/tav.yml also uses this file, so will need to be updated if this format changes.

File: .ci/tav.json [hunk 2]
```diff
@@ -1,39 +1,38 @@
 {
-  "// todo": "We want versions=['20','19','18','16','14','12','10','8'], but versions*modules needs to be <256 for the GH Actions jobs limit",
-  "versions": [ "20", "18", "16", "14", "12", "10", "8" ],
+  "versions": [ 20, 19, 18, 16, 14 ],
```
Comment (trentm) [inline .ci/tav.json:2]: If we want for 3.x, then we'll want 12, 10, and 8 back.

File: .github/workflows/tav-command.yml [hunk 3]
```diff
@@ -72,15 +72,38 @@ jobs:
...
+            const badVersionIndex = versions.findIndex((v) => isNaN(v))
+
+            if (badModuleIndex !== -1) {
+              core.setFailed(`Incorrect modules list, please review it'`)
+              return
+            }
+            if (badVersionIndex !== -1) {
+              core.setFailed(`Incorrect versions list, please review it'`)
+              return
+            }
```
Comment (trentm) [inline .github/workflows/tav-command.yml:93]: nit: I think the processing here could be a little more helpful and include the unknown module or version in the error message. Also it could perhaps handle *multiple* unknown modules and versions.  `/test tav foo,bar,baz`

File: .github/workflows/tav-command.yml [hunk 4]
```diff
@@ -72,15 +72,38 @@ jobs:
...
+
+            const permutations = []
+            for (const mod of modules) {
+              for (const nv of versions) {
+                if (mod.minVersion && nv >= mod.minVersion) {
+                  permutations.push(`${mod.name} ${nv}`)
+                } else if (typeof mod === 'string') {
```
Comment (trentm) [inline .github/workflows/tav-command.yml:100]: When can `mod` be a string? I think the processing above ensures it is always an array of entries like `{"name": "tedious", "minVersion": 18 }`.

File: .github/workflows/tav-command.yml [hunk 5]
```diff
@@ -72,15 +72,38 @@ jobs:
...
+                  permutations.push(`${mod} ${nv}`)
+                }
+              }
+            }
+            core.setOutput('permutations', permutations)
+            core.setOutput('count', [permutations.length])
```
Comment (trentm) [inline .github/workflows/tav-command.yml:106]: As we discussed, it would be nice to remove `count` so it isn't in the job names.

Would it be possible to error out or warn if `permutations.length > 256`?

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -134,13 +134,6 @@ jobs:
           - '16.0'
           - '14'
           - '14.0'
-          - '12'
```
Comment (trentm) [inline .github/workflows/test.yml:137]: Are some of these changes because you originally used some of the commits from `dev/4.x` and then retargetted main?

File: .github/workflows/tav-command.yml (hunk 4 above)
Comment (david-luna) [reply .github/workflows/tav-command.yml:100]: Leftover from a previous iteration where the `modules.findIndex` was not used

File: .github/workflows/tav-command.yml (hunk 5 above)
Comment (david-luna) [reply .github/workflows/tav-command.yml:106]: > Would it be possible to error out or warn if permutations.length > 256?

That's a fantastic idea :)

File: .github/workflows/test.yml (hunk 6 above)
Comment (david-luna) [reply .github/workflows/test.yml:137]: That's right. I'm going to add them again

**Review by david-luna — COMMENTED** [review]
/test tav express

**Review by david-luna — COMMENTED** [review]
/test tav express,fastify 20

**Comment by david-luna** [comment]
Turns out you can pass a single array (so it's a 1 dimension matrix). Now the list of the tasks looks nicer showing.the module name and the node version separated by a space.

Screenshot 2023-08-02 at 12 42 08

**Review by david-luna — COMMENTED** [review]
/test tav express,nonExistingPackage,fastify

**Comment by david-luna** [comment]
> /test tav express,nonExistingPackage,fastify

Now it fails and logs the name of the wrong package
Screenshot 2023-08-02 at 12 50 19

File: .ci/tav.json
Comment (david-luna) [reply .ci/tav.json:1]: Thanks for the heads up! I've changed it accordingly

**Review by david-luna — COMMENTED** [review]
/test tav

**Review by david-luna — COMMENTED** [review]
/test tav

File: .ci/tav.json [hunk 7]
```diff
@@ -1,39 +1,38 @@
...
+    {"name": "next", "minVersion": 14 },
+    {"name": "pg", "minVersion": 8 },
+    {"name": "redis", "minVersion": 8 },
+    {"name": "restify", "minVersion": 14 },
+    {"name": "tedious", "minVersion": 8 },
+    {"name": "undici", "minVersion": 8 },
+    {"name": "ws", "minVersion": 8 },
+    {"name": "@koa/router,koa-router", "minVersion": 8 },
+    {"name": "handlebars,pug", "minVersion": 8 },
+    {"name": "bluebird,got", "minVersion": 8 }
```
Comment (trentm) [inline .ci/tav.json:36]: Hrm, I guess with all these `minVersion: 8`s, it means there won't be a large reduction in the number of jobs. Still probably worth keeping.

(Not part of this PR, but I wonder if we could consider dropping testing for versions of these packages that are, say, older than 2y and not of the latest major version. That might enable raising the `minVersion` here.)

File: .ci/tav.json [hunk 8]
```diff
@@ -1,39 +1,38 @@
 {
-  "// todo": "We want versions=['20','19','18','16','14','12','10','8'], but versions*modules needs to be <256 for the GH Actions jobs limit",
-  "versions": [ "20", "18", "16", "14", "12", "10", "8" ],
+  "versions": [ 20, 19, 18, 16, 14, 12, 10, 8 ],
   "modules": [
```
Comment (trentm) [inline .ci/tav.json:3]: Might be nice to have a `"// modules": "..."` comment that explains minVersion and suggests it should be kept in sync with .tav.yml.

File: .github/workflows/tav-command.yml [hunk 9]
```diff
@@ -55,32 +54,80 @@ jobs:
...
+              if (match[1] === 'all') {
+                resolvedModules.push(...modules)
+              } else {
+                inputNames = match[1].split(',')
+                for (const name of inputNames) {
+                  const mod = modules.find((m) => m.name === name)
+                  if (mod) {
+                    resolvedModules.push(mod)
+                  } else {
+                    core.setFailed(`Incorrect module name ${name}, please review it'`)
```
Comment (trentm) [inline .github/workflows/tav-command.yml:108]: nit:
```suggestion
                    core.setFailed(`Incorrect module name ${name}, please review it`)
```

File: .github/workflows/tav-command.yml [hunk 10]
```diff
@@ -55,32 +54,80 @@ jobs:
...
+                  }
                 }
               }
-              if (match[3]) {
-                versions = match[3].split(',')
+            }
+            if (match[3]) {
+              inputVersions = match[3].split(',').map(Number)
+              if (inputVersions.some((v) => isNaN(v))) {
+                core.setFailed(`Incorrect versions list ${match[3]}, please review it'`)
```
Comment (trentm) [inline .github/workflows/tav-command.yml:117]: nit:
```suggestion
                core.setFailed(`Incorrect versions list ${match[3]}, please review it`)
```

File: dev-utils/lint-tav-json.js [hunk 11]
```diff
@@ -35,33 +35,43 @@ function main(argv) {
...
 
   // Matrix 256 limit.
-  const matrixSize = tavJson.versions.length * tavJson.modules.length;
-  if (matrixSize > 256) {
+  if (matrix.length > 256) {
     console.error(
       'lint-tav-json: #versions * #modules from "%s" is >256, which exceeds the GH Actions workflow matrix limit: %d',
       TAV_JSON_PATH,
-      matrixSize,
+      matrix.length,
```
Comment (trentm) [inline dev-utils/lint-tav-json.js:74]: Oh, I had forgotten that we had linting for being >256 already.

**Review by trentm — APPROVED** [review]
LGTM. A couple of nits is all.

File: .ci/tav.json (hunk 7 above)
Comment (david-luna) [reply .ci/tav.json:36]: Maybe we can leverage the multiple package option we have, here we are testing 2 packages in the same job
```
{"name": "bluebird,got", "minVersion": 8 }
```
This is actually what I want to do for `@aws-sdk/client-*` to test all of them at once for each version. There might be other groups we could make perhaps `@opentelemetry/*` or `mysql`? Of course some test review is needed to see compatibility. Also I think the `runTestFixtures` util introduced in the [ESM support PR](https://github.com/elastic/apm-agent-nodejs/pull/3381/files#diff-4044438594f145ca4e5b9920e844237cce63dd56744a23e3f263456e6a9f7d35R246) would come handy

File: dev-utils/lint-tav-json.js (hunk 11 above)
Comment (david-luna) [reply dev-utils/lint-tav-json.js:74]: Which is really good :)

File: .ci/tav.json (hunk 8 above)
Comment (david-luna) [reply .ci/tav.json:7]: Done. We may also want this in the TAV linting 🤔

**Review by david-luna — COMMENTED** [review]
/test tav fastify 20
